lib/utils/sortcache/sortcache.go

```diff
@@ -0,0 +1,234 @@
+	rw      sync.RWMutex
+	indexes map[string]func(T) string
+	trees   map[string]*btree.BTreeG[entry]
+	values  map[uint64]T
+	counter uint64
+}
+
+type entry struct {
+	key string
+	ref uint64
+}
+
+// New sets up a new [SortCache] based on the provided configuration.
+func New[T any](cfg Config[T]) *SortCache[T] {
+	const (
+		// bTreeDegree of 8 is standard across most of the teleport codebase
+		bTreeDegree = 8
+	)
+
+	trees := make(map[string]*btree.BTreeG[entry], len(cfg.Indexes))
+
+	for index := range cfg.Indexes {
+		trees[index] = btree.NewG(bTreeDegree, func(a, b entry) bool {
+			return a.key < b.key
+		})
+	}
+
+	return &SortCache[T]{
+		indexes: cfg.Indexes,
+		trees:   trees,
+		values:  make(map[uint64]T),
+	}
+}
+
+// Get loads the value associated with the given index and key. ok will be false if either
+// the index does not exist or no value maps to the provided key on that index. Note that
+// mutating a value such that any of its index keys change is not permissible and will result
+// in permanently bad state. To avoid this, any implementation that might mutate returned
+// values must clone them.
+func (c *SortCache[T]) Get(index, key string) (value T, ok bool) {
+	c.rw.RLock()
+	defer c.rw.RUnlock()
+
+	ref, ok := c.lookup(index, key)
+	if !ok {
+		return value, false
+	}
+
+	return c.values[ref], true
+}
+
+// HasIndex checks if the specified index is present in this cache.
+func (c *SortCache[T]) HasIndex(name string) bool {
+	// index map is treated as immutable once created, so no lock is required.
+	_, ok := c.indexes[name]
+	return ok
+}
+
+// KeyOf gets the key of the supplied value on the given index.
+func (c *SortCache[T]) KeyOf(index string, value T) string {
+	// index map is treated as immutable once created, so no lock is required.
+	fn, ok := c.indexes[index]
+	if !ok {
+		return ""
+	}
+	return fn(value)
+}
+
+// lookup is an internal helper that finds the unique reference id for a value given a specific
+// index/key. Must be called either under read or write lock.
+func (c *SortCache[T]) lookup(index, key string) (ref uint64, ok bool) {
+	tree, exists := c.trees[index]
+	if !exists {
+		return 0, false
+	}
+
+	entry, exists := tree.Get(entry{key: key})
+	if !exists {
+		return 0, false
+	}
+
+	return entry.ref, true
+}
+
+// Put inserts a value into the sort cache, removing any existing values that collide with it. Since all indexes
+// are required to be unique, a single Put can end up evicting up to N existing values where N is the number of
+// indexes if it happens to collide with a different value on each index. implementations that expect evictions
+// to always either be zero or one (e.g. caches of resources with unique IDs) should check the returned eviction
+// count to help detect bugs.
+func (c *SortCache[T]) Put(value T) (evicted int) {
+	c.rw.Lock()
+	defer c.rw.Unlock()
+
+	c.counter++
+	c.values[c.counter] = value
+
+	for index, fn := range c.indexes {
+		key := fn(value)
```

Comment:
would it make sense to add some sort of validation to the key here? maybe something as simple as "make sure it isn't empty string"

Reply:
Tried out some variations of this, but ended up reverting them.  The sortcache itself doesn't care if you use empty keys (or any particular kind of key), and its actually pretty tricky to do this logic in such a way where it can fail mid insertion without leaving the cache in a weird state (at least not without pre-allocating all the keys in a temporary mapping, which would be unfortunate for performance).

I think instead, users of this helper that have specific requirements for their keys (e.g. no empty keys), should use the `SortCache.KeyOf(...)` method to see what keys their values map to and decide if they are acceptable before insertion.  For the access-request usecase I'm not really worried about that (all indexes use the uuid as a suffix, and uuids are required to be non-empty already), and I'd therefore prefer not to pay the cost in extra allocations.